File: app/local_plugins/vendors/mcp/android/src/main/kotlin/com/aiagent/mcp/McpStreamableHttpService.kt
package com.aiagent.mcp

import android.util.Log
import com.aiagent.plugin_interface.NativeMcpService
import kotlinx.coroutines.suspendCancellableCoroutine
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * McpStreamableHttpService — MCP Streamable HTTP transport（MCP 2024-11-05）
 *
 * 协议：单一 endpoint，POST 发 JSON-RPC 2.0；响应 Content-Type 为
 * `application/json` 或 `text/event-stream`，后者按 SSE 取首个 `data:` 帧。
 * 服务器会在 `initialize` 响应中下发 `Mcp-Session-Id`，后续请求必须回带。
 *
 * 单实例对应单 server。多 server 由 `McpRouter` 聚合。
 */
class McpStreamableHttpService : NativeMcpService {

    companion object {
        private const val TAG = "McpStreamableHttp"
        private const val PROTOCOL_VERSION = "2024-11-05"
        private val JSON_MEDIA = "application/json".toMediaType()
    }

    @Volatile private var client: OkHttpClient? = null
    private var url: String = ""
    private var authHeader: String? = null
    private val extraHeaders = mutableMapOf<String, String>()
    private var enabledTools: Set<String> = emptySet()
    private var serverId: String = ""

    @Volatile private var sessionId: String? = null
    @Volatile private var initialized: Boolean = false
    @Volatile private var rpcId: Int = 0

    // ─────────────────────────────────────────────────
    // NativeMcpService
    // ─────────────────────────────────────────────────

    override suspend fun initialize(configJson: String) {
        dispose()

        val cfg = JSONObject(configJson)
        serverId = cfg.optString("id", "")
        url = cfg.optString("url", "").also {
            require(it.isNotBlank()) { "McpServerConfig.url required" }
        }
        authHeader = cfg.optString("authHeader", "").ifBlank { null }
        val timeoutSec = cfg.optInt("timeoutSeconds", 30).coerceAtLeast(1)

        enabledTools = cfg.optJSONArray("enabledTools")?.let { arr ->
            buildSet {
                for (i in 0 until arr.length()) add(arr.optString(i))
            }
        } ?: emptySet()

        cfg.optJSONObject("extraHeaders")?.let { extras ->
            for (k in extras.keys()) extraHeaders[k] = extras.optString(k, "")
        }

        client = OkHttpClient.Builder()
            .connectTimeout(timeoutSec.toLong(), TimeUnit.SECONDS)
            .readTimeout(timeoutSec.toLong(), TimeUnit.SECONDS)
            .writeTimeout(timeoutSec.toLong(), TimeUnit.SECONDS)
            .build()

        // JSON-RPC initialize 握手
        val initParams = JSONObject().apply {
            put("protocolVersion", PROTOCOL_VERSION)
            put("capabilities", JSONObject().apply { put("tools", JSONObject()) })
            put("clientInfo", JSONObject().apply {
                put("name", "ai-agent-client")
                put("version", "1.0.0")
            })
        }
        val initResult = rpc("initialize", initParams)
        if (!initResult.has("protocolVersion") && !initResult.has("serverInfo")) {
            throw IllegalStateException("mcp.handshake_failed: initialize 响应缺少 protocolVersion / serverInfo")
        }
        rpcNotification("notifications/initialized", JSONObject())
        initialized = true
        Log.d(TAG, "Initialized server: $serverId @ $url")
    }

    override suspend fun listTools(): List<Map<String, Any?>> {
        ensureReady()
        val result = rpc("tools/list", JSONObject())
        val arr = result.optJSONArray("tools") ?: return emptyList()
        val out = mutableListOf<Map<String, Any?>>()
        for (i in 0 until arr.length()) {
            val t = arr.optJSONObject(i) ?: continue
            val name = t.optString("name", "")
            if (name.isBlank()) continue
            if (enabledTools.isNotEmpty() && name !in enabledTools) continue
            val schema = t.optJSONObject("inputSchema")?.let { jsonObjectToMap(it) } ?: emptyMap<String, Any?>()
            out.add(
                mapOf(
                    "name" to name,
                    "description" to t.optString("description", ""),
                    "inputSchema" to schema,
                    "serverId" to serverId,
                )
            )
        }
        return out
    }

    override suspend fun callTool(toolName: String, argsJson: String): Map<String, Any?> {
        ensureReady()
        val argsObj = if (argsJson.isBlank()) JSONObject() else try {
            JSONObject(argsJson)
        } catch (_: Exception) {
            JSONObject()
        }
        val params = JSONObject().apply {
            put("name", toolName)
            put("arguments", argsObj)
        }
        val result = rpc("tools/call", params)
        val isError = result.optBoolean("isError", false)
        val contentArr = result.optJSONArray("content")
        val sb = StringBuilder()
        if (contentArr != null) {
            for (i in 0 until contentArr.length()) {
                val c = contentArr.optJSONObject(i) ?: continue
                if (c.optString("type") == "text") {
                    val txt = c.optString("text", "")
                    if (txt.isNotEmpty()) {
                        if (sb.isNotEmpty()) sb.append("\n")
                        sb.append(txt)
                    }
                }
            }
        }
        return mapOf("content" to sb.toString(), "isError" to isError)
    }

    override fun dispose() {
        client = null
        sessionId = null
        initialized = false
        rpcId = 0
        enabledTools = emptySet()
        extraHeaders.clear()
        authHeader = null
    }

    // ─────────────────────────────────────────────────
    // 内部
    // ─────────────────────────────────────────────────

    private fun ensureReady() {
        check(client != null) { "McpStreamableHttpService disposed" }
        check(initialized) { "McpStreamableHttpService not initialized — call initialize() first" }
    }

    private fun buildHeaders(builder: Request.Builder): Request.Builder {
        builder.header("Content-Type", "application/json")
        builder.header("Accept", "application/json, text/event-stream")
        authHeader?.takeIf { it.isNotBlank() }?.let { builder.header("Authorization", it) }
        sessionId?.takeIf { it.isNotBlank() }?.let { builder.header("Mcp-Session-Id", it) }
        for ((k, v) in extraHeaders) builder.header(k, v)
        return builder
    }

    private suspend fun rpc(method: String, params: JSONObject): JSONObject {
        rpcId += 1
        val body = JSONObject().apply {
            put("jsonrpc", "2.0")
            put("id", rpcId)
            put("method", method)
            put("params", params)
        }
        val resp = postWithBody(body)
        if (resp.statusCode != 200 && resp.statusCode != 202) {
            throw IllegalStateException("mcp.http_${resp.statusCode}: ${resp.bodySnippet()}")
        }
        val rpcObj = parseRpcBody(resp)
        rpcObj.optJSONObject("error")?.let { err ->
            throw IllegalStateException("mcp.rpc_${err.optInt("code")}: ${err.optString("message")}")
        }
        return rpcObj.optJSONObject("result") ?: JSONObject()
    }

    private suspend fun rpcNotification(method: String, params: JSONObject) {
        val body = JSONObject().apply {
            put("jsonrpc", "2.0")
            put("method", method)
            put("params", params)
        }
        val resp = postWithBody(body)
        if (resp.statusCode >= 400) {
            throw IllegalStateException("mcp.http_${resp.statusCode}: ${resp.bodySnippet()}")
        }
    }

    private suspend fun postWithBody(body: JSONObject): _RpcResponse {
        val req = Request.Builder().url(url).also { buildHeaders(it) }
            .post(body.toString().toRequestBody(JSON_MEDIA))
            .build()
        val response = enqueue(req)
        try {
            response.header("Mcp-Session-Id")?.takeIf { it.isNotBlank() }?.let { sessionId = it }
            val ct = response.header("Content-Type") ?: ""
            val text = response.body?.string() ?: ""
            return _RpcResponse(response.code, ct, text)
        } finally {
            response.close()
        }
    }

    private suspend fun enqueue(request: Request): Response {
        val c = client ?: throw IllegalStateException("McpStreamableHttpService disposed")
        return suspendCancellableCoroutine { cont ->
            val call = c.newCall(request)
            cont.invokeOnCancellation { call.cancel() }
            call.enqueue(object : Callback {
                override fun onResponse(call: Call, response: Response) {
                    if (cont.isActive) cont.resumeWith(Result.success(response))
                    else response.close()
                }
                override fun onFailure(call: Call, e: IOException) {
                    if (cont.isActive) cont.resumeWith(Result.failure(e))
                }
            })
        }
    }

    private fun parseRpcBody(resp: _RpcResponse): JSONObject {
        if (resp.contentType.contains("text/event-stream", ignoreCase = true)) {
            for (raw in resp.body.split('\n')) {
                val line = raw.trimEnd()
                if (line.startsWith("data:")) {
                    val payload = line.substring(5).trimStart()
                    if (payload.isEmpty()) continue
                    return JSONObject(payload)
                }
            }
            throw IllegalStateException("mcp.invalid_response: SSE 响应未包含 JSON-RPC data 帧")
        }
        if (resp.body.isBlank()) return JSONObject()
        return JSONObject(resp.body)
    }

    private fun jsonObjectToMap(obj: JSONObject): Map<String, Any?> {
        val out = mutableMapOf<String, Any?>()
        for (k in obj.keys()) out[k] = jsonValueToAny(obj.get(k))
        return out
    }

    private fun jsonArrayToList(arr: JSONArray): List<Any?> {
        val out = mutableListOf<Any?>()
        for (i in 0 until arr.length()) out.add(jsonValueToAny(arr.get(i)))
        return out
    }

    private fun jsonValueToAny(v: Any?): Any? = when (v) {
        is JSONObject -> jsonObjectToMap(v)
        is JSONArray -> jsonArrayToList(v)
        JSONObject.NULL -> null
        else -> v
    }

    private data class _RpcResponse(val statusCode: Int, val contentType: String, val body: String) {
        fun bodySnippet(): String = if (body.length > 200) body.substring(0, 200) + "…" else body
    }
}

File: apps/client/local_plugins/vendors/mcp/ios/Sources/mcp/McpStreamableHttpService.swift
import Foundation

/// McpStreamableHttpService — MCP Streamable HTTP transport（MCP 2024-11-05）（iOS 端）
///
/// 协议要点：
///   - 单一 endpoint，所有方法走 POST
///   - 请求 body 为 JSON-RPC 2.0
///   - 响应 Content-Type 可能是 `application/json` 或 `text/event-stream`
///   - 会话头：`Mcp-Session-Id`，server 在 `initialize` 响应中返回，后续请求必须回带
///
/// 本文件目前**未启用**：iOS native 链路尚未铺开（参见 `NativeMcpService.swift`
/// 顶部说明）。当 iOS 端 chat agent 链路整体启动时，把它注册进 iOS 版的
/// `NativeServiceRegistry`（注册键 `streamable_http`）即可对接。
///
/// 单实例对应单 server。多 server 聚合应由 iOS 版 `NativeMcpRouter` 负责。
public final class McpStreamableHttpService: NativeMcpService {

    private static let protocolVersion = "2024-11-05"

    private var session: URLSession?
    private var url: URL?
    private var authHeader: String?
    private var extraHeaders: [String: String] = [:]
    private var enabledTools: Set<String> = []
    private var serverId: String = ""
    private var timeoutSeconds: TimeInterval = 30

    private var sessionId: String?
    private var initialized: Bool = false
    private var rpcId: Int = 0

    public init() {}

    // MARK: - NativeMcpService

    public func initialize(configJson: String) async throws {
        dispose()

        guard let data = configJson.data(using: .utf8),
              let cfg = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            throw NSError(domain: "mcp.invalid_config", code: -1)
        }

        serverId = cfg["id"] as? String ?? ""
        guard let urlStr = cfg["url"] as? String, !urlStr.isEmpty,
              let parsed = URL(string: urlStr)
        else {
            throw NSError(domain: "mcp.invalid_config", code: -1, userInfo: [NSLocalizedDescriptionKey: "url required"])
        }
        self.url = parsed
        self.authHeader = (cfg["authHeader"] as? String).flatMap { $0.isEmpty ? nil : $0 }
        self.timeoutSeconds = TimeInterval(cfg["timeoutSeconds"] as? Int ?? 30)
        self.enabledTools = Set((cfg["enabledTools"] as? [String]) ?? [])
        if let extras = cfg["extraHeaders"] as? [String: String] {
            self.extraHeaders = extras
        }

        let sessConfig = URLSessionConfiguration.default
        sessConfig.timeoutIntervalForRequest = timeoutSeconds
        sessConfig.timeoutIntervalForResource = timeoutSeconds
        self.session = URLSession(configuration: sessConfig)

        // JSON-RPC initialize 握手
        let initParams: [String: Any] = [
            "protocolVersion": Self.protocolVersion,
            "capabilities": ["tools": [String: Any]()],
            "clientInfo": ["name": "ai-agent-client", "version": "1.0.0"],
        ]
        let initResult = try await rpc("initialize", params: initParams)
        if initResult["protocolVersion"] == nil && initResult["serverInfo"] == nil {
            throw NSError(domain: "mcp.handshake_failed", code: -1, userInfo: [NSLocalizedDescriptionKey: "initialize 响应缺少 protocolVersion / serverInfo"])
        }
        try await rpcNotification("notifications/initialized", params: [String: Any]())
        initialized = true
    }

    public func listTools() async throws -> [[String: Any?]] {
        try ensureReady()
        let result = try await rpc("tools/list", params: [String: Any]())
        guard let arr = result["tools"] as? [[String: Any]] else { return [] }
        var out: [[String: Any?]] = []
        for t in arr {
            guard let name = t["name"] as? String, !name.isEmpty else { continue }
            if !enabledTools.isEmpty && !enabledTools.contains(name) { continue }
            out.append([
                "name": name,
                "description": (t["description"] as? String) ?? "",
                "inputSchema": (t["inputSchema"] as? [String: Any]) ?? [:],
                "serverId": serverId,
            ])
        }
        return out
    }

    public func callTool(toolName: String, argsJson: String) async throws -> [String: Any?] {
        try ensureReady()
        let argsObj: [String: Any]
        if argsJson.isEmpty {
            argsObj = [:]
        } else if let data = argsJson.data(using: .utf8),
                  let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            argsObj = parsed
        } else {
            argsObj = [:]
        }
        let result = try await rpc("tools/call", params: ["name": toolName, "arguments": argsObj])
        let isError = (result["isError"] as? Bool) ?? false
        var text = ""
        if let arr = result["content"] as? [[String: Any]] {
            for c in arr where (c["type"] as? String) == "text" {
                if let t = c["text"] as? String, !t.isEmpty {
                    if !text.isEmpty { text += "\n" }
                    text += t
                }
            }
        }
        return ["content": text, "isError": isError]
    }

    public func dispose() {
        session?.invalidateAndCancel()
        session = nil
        url = nil
        authHeader = nil
        extraHeaders.removeAll()
        enabledTools.removeAll()
        sessionId = nil
        initialized = false
        rpcId = 0
    }

    // MARK: - Internal

    private func ensureReady() throws {
        guard session != nil else {
            throw NSError(domain: "mcp.disposed", code: -1)
        }
        guard initialized else {
            throw NSError(domain: "mcp.not_initialized", code: -1, userInfo: [NSLocalizedDescriptionKey: "call initialize() first"])
        }
    }

    private func buildRequest(body: Data) throws -> URLRequest {
        guard let url = url else { throw NSError(domain: "mcp.disposed", code: -1) }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json, text/event-stream", forHTTPHeaderField: "Accept")
        if let auth = authHeader { req.setValue(auth, forHTTPHeaderField: "Authorization") }
        if let sid = sessionId { req.setValue(sid, forHTTPHeaderField: "Mcp-Session-Id") }
        for (k, v) in extraHeaders { req.setValue(v, forHTTPHeaderField: k) }
        req.httpBody = body
        return req
    }

    private func rpc(_ method: String, params: [String: Any]) async throws -> [String: Any] {
        rpcId += 1
        let body: [String: Any] = [
            "jsonrpc": "2.0",
            "id": rpcId,
            "method": method,
            "params": params,
        ]
        let data = try JSONSerialization.data(withJSONObject: body)
        let req = try buildRequest(body: data)
        guard let session = session else { throw NSError(domain: "mcp.disposed", code: -1) }
        let (respData, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse else {
            throw NSError(domain: "mcp.invalid_response", code: -1)
        }
        if let sid = http.value(forHTTPHeaderField: "Mcp-Session-Id"), !sid.isEmpty {
            sessionId = sid
        }
        if http.statusCode != 200 && http.statusCode != 202 {
            let snippet = String(data: respData, encoding: .utf8)?.prefix(200) ?? ""
            throw NSError(domain: "mcp.http_\(http.statusCode)", code: http.statusCode, userInfo: [NSLocalizedDescriptionKey: String(snippet)])
        }
        let rpcObj = try parseRpcBody(data: respData, contentType: http.value(forHTTPHeaderField: "Content-Type") ?? "")
        if let err = rpcObj["error"] as? [String: Any] {
            let code = err["code"] as? Int ?? -1
            let msg = err["message"] as? String ?? ""
            throw NSError(domain: "mcp.rpc_\(code)", code: code, userInfo: [NSLocalizedDescriptionKey: msg])
        }
        return (rpcObj["result"] as? [String: Any]) ?? [:]
    }

    private func rpcNotification(_ method: String, params: [String: Any]) async throws {
        let body: [String: Any] = [
            "jsonrpc": "2.0",
            "method": method,
            "params": params,
        ]
        let data = try JSONSerialization.data(withJSONObject: body)
        let req = try buildRequest(body: data)
        guard let session = session else { throw NSError(domain: "mcp.disposed", code: -1) }
        let (_, response) = try await session.data(for: req)
        if let http = response as? HTTPURLResponse,
           let sid = http.value(forHTTPHeaderField: "Mcp-Session-Id"), !sid.isEmpty {
            sessionId = sid
        }
        if let http = response as? HTTPURLResponse, http.statusCode >= 400 {
            throw NSError(domain: "mcp.http_\(http.statusCode)", code: http.statusCode)
        }
    }

    private func parseRpcBody(data: Data, contentType: String) throws -> [String: Any] {
        if contentType.lowercased().contains("text/event-stream") {
            let body = String(data: data, encoding: .utf8) ?? ""
            for raw in body.split(separator: "\n") {
                let line = raw.trimmingCharacters(in: .whitespacesAndNewlines)
                if line.hasPrefix("data:") {
                    let payload = String(line.dropFirst(5)).trimmingCharacters(in: .whitespaces)
                    if payload.isEmpty { continue }
                    if let pdata = payload.data(using: .utf8),
                       let obj = try JSONSerialization.jsonObject(with: pdata) as? [String: Any] {
                        return obj
                    }
                }
            }
            throw NSError(domain: "mcp.invalid_response", code: -1, userInfo: [NSLocalizedDescriptionKey: "SSE 响应未包含 JSON-RPC data 帧"])
        }
        if data.isEmpty { return [:] }
        guard let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw NSError(domain: "mcp.invalid_response", code: -1)
        }
        return obj
    }
}
